Importing numpy and Pandas package

In [1]:
import numpy as np
import pandas as pd

Reading CSV and making DataFrame as policedf

In [2]:
policedf = pd.read_csv('PoliceData.csv') 

Checking top 5 records in dataframe using head() function

In [3]:
policedf.head()

stop_date stop_time  country_name driver_gender  driver_age_raw  \
0   1/2/2005      1:55           NaN             M          1985.0   
1  1/18/2005      8:15           NaN             M          1965.0   
2  1/23/2005     23:15           NaN             M          1972.0   
3  2/20/2005     17:15           NaN             M          1986.0   
4  3/14/2005     10:00           NaN             F          1984.0   

   driver_age driver_race     violation_raw violation  search_conducted  \
0        20.0       White          Speeding  Speeding             False   
1        40.0       White          Speeding  Speeding             False   
2        33.0       White          Speeding  Speeding             False   
3        19.0       White  Call for Service     Other             False   
4        21.0       White          Speeding  Speeding             False   

  search_type   stop_outcome is_arrested stop_duration  drugs_related_stop  
0         NaN       Citation       False      0-15 Min               False  
1         NaN       Citation       False      0-15 Min               False  
2         NaN       Citation       False      0-15 Min               False  
3         NaN  Arrest Driver        True     16-30 Min               False  
4         NaN       Citation       False      0-15 Min               False

Checking shape of the data frame. Means Rows and Columns

In [4]:
policedf.shape

(65535, 15)

Setting Pandas option to view all columns in our Jupyter Notebook

In [5]:
pd.set_option('display.max_columns', 15)

Getting information on each column using info() function

In [6]:
policedf.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65535 entries, 0 to 65534
Data columns (total 15 columns):
stop_date             65535 non-null object
stop_time             65535 non-null object
country_name          0 non-null float64
driver_gender         61474 non-null object
driver_age_raw        61481 non-null float64
driver_age            61228 non-null float64
driver_race           61475 non-null object
violation_raw         61475 non-null object
violation             61475 non-null object
search_conducted      65535 non-null bool
search_type           2479 non-null object
stop_outcome          61475 non-null object
is_arrested           61475 non-null object
stop_duration         61475 non-null object
drugs_related_stop    65535 non-null bool
dtypes: bool(2), float64(3), object(10)
memory usage: 6.6+ MB


Using describe function to get more detailed info on our Dataset

In [7]:
policedf.describe(include='all')

stop_date stop_time  country_name driver_gender  driver_age_raw  \
count       65535     65535           0.0         61474    61481.000000   
unique       2651      1432           NaN             2             NaN   
top     1/11/2012     10:00           NaN             M             NaN   
freq           69       280           NaN         45164             NaN   
mean          NaN       NaN           NaN           NaN     1967.791106   
std           NaN       NaN           NaN           NaN      121.050106   
min           NaN       NaN           NaN           NaN        0.000000   
25%           NaN       NaN           NaN           NaN     1965.000000   
50%           NaN       NaN           NaN           NaN     1978.000000   
75%           NaN       NaN           NaN           NaN     1985.000000   
max           NaN       NaN           NaN           NaN     8801.000000   

          driver_age driver_race violation_raw violation search_conducted  \
count   61228.000000       61475         61475     61475            65535   
unique           NaN           5            12         6                2   
top              NaN       White      Speeding  Speeding            False   
freq             NaN       45747         37204     37204            63056   
mean       34.148984         NaN           NaN       NaN              NaN   
std        12.760710         NaN           NaN       NaN              NaN   
min        15.000000         NaN           NaN       NaN              NaN   
25%        23.000000         NaN           NaN       NaN              NaN   
50%        31.000000         NaN           NaN       NaN              NaN   
75%        43.000000         NaN           NaN       NaN              NaN   
max        88.000000         NaN           NaN       NaN              NaN   

               search_type stop_outcome is_arrested stop_duration  \
count                 2479        61475       61475         61475   
unique                  23            6           2             4   
top     Incident to Arrest     Citation       False      0-15 Min   
freq                  1074        56564       59215         47379   
mean                   NaN          NaN         NaN           NaN   
std                    NaN          NaN         NaN           NaN   
min                    NaN          NaN         NaN           NaN   
25%                    NaN          NaN         NaN           NaN   
50%                    NaN          NaN         NaN           NaN   
75%                    NaN          NaN         NaN           NaN   
max                    NaN          NaN         NaN           NaN   

       drugs_related_stop  
count               65535  
unique                  2  
top                 False  
freq                65017  
mean                  NaN  
std                   NaN  
min                   NaN  
25%                   NaN  
50%                   NaN  
75%                   NaN  
max                   NaN

Checking the columns in dataframe using columns attribute

In [8]:
policedf.columns

Index(['stop_date', 'stop_time', 'country_name', 'driver_gender',
       'driver_age_raw', 'driver_age', 'driver_race', 'violation_raw',
       'violation', 'search_conducted', 'search_type', 'stop_outcome',
       'is_arrested', 'stop_duration', 'drugs_related_stop'],
      dtype='object')

counting the number rows null in each column

In [9]:
policedf.isnull().sum()

stop_date                 0
stop_time                 0
country_name          65535
driver_gender          4061
driver_age_raw         4054
driver_age             4307
driver_race            4060
violation_raw          4060
violation              4060
search_conducted          0
search_type           63056
stop_outcome           4060
is_arrested            4060
stop_duration          4060
drugs_related_stop        0
dtype: int64

As we can see from above details that
columns 'country_name' and 'search_type' does have maximum values as null
hench we will drop those columns

In [10]:
policedf.drop(['country_name', 'search_type'], axis=1, inplace=True)

In [11]:
policedf.head(20)

stop_date stop_time driver_gender  driver_age_raw  driver_age driver_race  \
0    1/2/2005      1:55             M          1985.0        20.0       White   
1   1/18/2005      8:15             M          1965.0        40.0       White   
2   1/23/2005     23:15             M          1972.0        33.0       White   
3   2/20/2005     17:15             M          1986.0        19.0       White   
4   3/14/2005     10:00             F          1984.0        21.0       White   
5   3/23/2005      9:45             M          1982.0        23.0       Black   
6    4/1/2005     17:30             M          1969.0        36.0       White   
7    6/6/2005     13:20             F          1986.0        19.0       White   
8   7/13/2005     10:15             M          1970.0        35.0       Black   
9   7/13/2005     15:45             M          1970.0        35.0       White   
10  7/13/2005     16:20             M          1979.0        26.0       Asian   
11  7/13/2005     19:00             F          1966.0        39.0       White   
12  7/14/2005     19:55             M          1979.0        26.0       White   
13  7/18/2005     19:30             F          1984.0        21.0       White   
14  7/18/2005     19:45             M          1969.0        36.0       White   
15  7/19/2005      0:30             M          1982.0        23.0       White   
16  7/19/2005      0:30             M          1982.0        23.0       White   
17  7/19/2005     23:30             M          1979.0        26.0       White   
18  7/20/2005      0:05             M          1955.0        50.0       White   
19  7/24/2005     20:10             F          1958.0        47.0       White   

                     violation_raw  violation  search_conducted  \
0                         Speeding   Speeding             False   
1                         Speeding   Speeding             False   
2                         Speeding   Speeding             False   
3                 Call for Service      Other             False   
4                         Speeding   Speeding             False   
5   Equipment/Inspection Violation  Equipment             False   
6                         Speeding   Speeding             False   
7                         Speeding   Speeding             False   
8                         Speeding   Speeding             False   
9                         Speeding   Speeding             False   
10                        Speeding   Speeding             False   
11                        Speeding   Speeding             False   
12                        Speeding   Speeding             False   
13                        Speeding   Speeding             False   
14                        Speeding   Speeding             False   
15                        Speeding   Speeding             False   
16                        Speeding   Speeding             False   
17  Equipment/Inspection Violation  Equipment             False   
18                        Speeding   Speeding             False   
19                        Speeding   Speeding             False   

     stop_outcome is_arrested stop_duration  drugs_related_stop  
0        Citation       False      0-15 Min               False  
1        Citation       False      0-15 Min               False  
2        Citation       False      0-15 Min               False  
3   Arrest Driver        True     16-30 Min               False  
4        Citation       False      0-15 Min               False  
5        Citation       False      0-15 Min               False  
6        Citation       False      0-15 Min               False  
7        Citation       False      0-15 Min               False  
8        Citation       False      0-15 Min               False  
9        Citation       False      0-15 Min               False  
10       Citation       False      0-15 Min               False  
11       Citation       False      0-15 Min               False  
12       Citation       False      0

And now we can't see country name and serach type columns

And now we will fill other null values, First we will check the rows which have all the values as null

In [12]:
policedf.shape

(65535, 13)

In [13]:
policedf.index[policedf.isnull().all(1)]

Int64Index([], dtype='int64')

As there are no such rows with all the columns as Null values so now we will check for each and every column

In [14]:
policedf.isnull().sum()

stop_date                0
stop_time                0
driver_gender         4061
driver_age_raw        4054
driver_age            4307
driver_race           4060
violation_raw         4060
violation             4060
search_conducted         0
stop_outcome          4060
is_arrested           4060
stop_duration         4060
drugs_related_stop       0
dtype: int64

In [15]:
policedf['driver_gender'].value_counts()

M    45164
F    16310
Name: driver_gender, dtype: int64

As 4061 people didn't mentioned their gender so we can mark them as 'Not Mentioned'

In [16]:
policedf['driver_gender'] = policedf['driver_gender'].fillna('Not Mentioned')

In [17]:
policedf['driver_age_raw'].unique()

array([1985., 1965., 1972., 1986., 1984., 1982., 1969., 1970., 1979.,
       1966., 1955., 1958., 1953., 1967.,   nan, 1964., 1977., 1959.,
       1973., 1947., 1978., 1962., 1981., 1988., 1983., 1961., 1957.,
       1971., 1950., 1954., 1960., 1976., 1974., 1968., 1938., 1980.,
       1963., 1975., 1987., 1948., 1956., 1944., 1935.,    0., 1952.,
       1945., 1946., 1941., 1942., 1951., 1949., 2005., 1939., 1943.,
       1936., 1924., 1937., 1929., 1930., 1931., 1923., 1940., 1933.,
       1989., 1928., 1932., 1990., 1926., 2006., 1927., 1997., 1934.,
       1887., 1925., 1918., 2004., 1991., 2002., 2007., 1992., 8801.,
       2001., 2003., 1998., 1922., 1993., 2008., 2009., 2919., 2000.,
       1994., 1995., 2011., 1996.])

As 4054 people didn't mentioned their age_raw that is their year of birth
we can mark nan values as 0

In [18]:
policedf['driver_age_raw'] = policedf['driver_age_raw'].fillna(0)

In [19]:
policedf['driver_age'].unique()

array([20., 40., 33., 19., 21., 23., 36., 35., 26., 39., 50., 47., 52.,
       38., nan, 41., 28., 46., 32., 58., 27., 43., 24., 17., 22., 44.,
       48., 34., 55., 51., 45., 29., 31., 37., 67., 25., 42., 30., 18.,
       57., 49., 61., 70., 53., 60., 59., 64., 63., 54., 56., 66., 62.,
       69., 81., 68., 76., 75., 74., 82., 65., 72., 16., 77., 73., 71.,
       80., 79., 78., 88., 15., 83., 86., 84., 85.])

As 4307 people didn't mentioned their age hence we will mark the na values as mean age

In [20]:
mean_age = policedf['driver_age'].mean()
policedf['driver_age'] = policedf['driver_age'].fillna(mean_age)

In [21]:
policedf['driver_race'].value_counts()

White       45747
Black        8267
Hispanic     5611
Asian        1639
Other         211
Name: driver_race, dtype: int64

More than 45K rocords are with White then we can say maximum population in our data set is white hence we can mark the rest na values as 'White'

In [22]:
policedf['driver_race'] = policedf['driver_race'].fillna('White')

In [23]:
policedf['violation_raw'].value_counts()

Speeding                            37204
Other Traffic Violation             11926
Equipment/Inspection Violation       6516
Special Detail/Directed Patrol       2383
Registration Violation               2243
Call for Service                      793
Violation of City/Town Ordinance      170
Motorist Assist/Courtesy              138
APB                                    51
Suspicious Person                      38
Warrant                                10
Seatbelt Violation                      3
Name: violation_raw, dtype: int64

In [24]:
policedf[policedf['violation_raw'].isnull()].head(10)

stop_date stop_time  driver_gender  driver_age_raw  driver_age  \
25     9/3/2005     16:02  Not Mentioned             0.0   34.148984   
29    9/26/2005      1:05  Not Mentioned             0.0   34.148984   
52    10/1/2005     13:09  Not Mentioned             0.0   34.148984   
110   10/3/2005     19:50  Not Mentioned             0.0   34.148984   
124   10/4/2005     11:43  Not Mentioned             0.0   34.148984   
150   10/5/2005     10:08  Not Mentioned             0.0   34.148984   
205   10/7/2005      9:09  Not Mentioned             0.0   34.148984   
226   10/8/2005      3:13  Not Mentioned          1968.0   37.000000   
314  10/11/2005     10:30  Not Mentioned             0.0   34.148984   
318  10/11/2005     15:20  Not Mentioned             0.0   34.148984   

    driver_race violation_raw violation  search_conducted stop_outcome  \
25        White           NaN       NaN             False          NaN   
29        White           NaN       NaN             False          NaN   
52        White           NaN       NaN             False          NaN   
110       White           NaN       NaN             False          NaN   
124       White           NaN       NaN             False          NaN   
150       White           NaN       NaN             False          NaN   
205       White           NaN       NaN             False          NaN   
226       White           NaN       NaN             False          NaN   
314       White           NaN       NaN             False          NaN   
318       White           NaN       NaN             False          NaN   

    is_arrested stop_duration  drugs_related_stop  
25          NaN           NaN               False  
29          NaN           NaN               False  
52          NaN           NaN               False  
110         NaN           NaN               False  
124         NaN           NaN               False  
150         NaN           NaN               False  
205         NaN           NaN               False  
226         NaN           NaN               False  
314         NaN           NaN               False  
318         NaN           NaN               False

As per our records maximum drivers are with Null values does have other columns as Null too so we can mark this column as not mentioned

In [25]:
policedf['violation_raw'] = policedf['violation_raw'].fillna('Not Mentioned')

In [26]:
policedf['violation'].value_counts()

Speeding               37204
Moving violation       11926
Equipment               6516
Other                   3583
Registration/plates     2243
Seat belt                  3
Name: violation, dtype: int64

Similar like Violation_raw column we can mention 'Not Mentioned' for this column too

In [27]:
policedf['violation'] = policedf['violation'].fillna('Not Mentioned')

In [28]:
policedf['stop_outcome'].value_counts()

Citation            56564
Arrest Driver        2009
Warning              1889
N/D                   423
No Action             339
Arrest Passenger      251
Name: stop_outcome, dtype: int64

As like above columns 4060 rows are there without anything mentioned
so we can mark this column as not mentioned

In [29]:
policedf['stop_outcome'] = policedf['stop_outcome'].fillna('Not Mentioned')

In [30]:
policedf['is_arrested'].value_counts()

False    59215
True      2260
Name: is_arrested, dtype: int64

Like above columns as most of the details are not there for driver hence we can mark na values as 'Not Mentioned'

In [31]:
policedf['is_arrested'] = policedf['is_arrested'].fillna('Not Mentioned')

In [32]:
policedf['stop_duration'].value_counts()

0-15 Min     47379
16-30 Min    11448
30+ Min       2647
2                1
Name: stop_duration, dtype: int64

Like above columns as most of the details are not there for driver hence we can mark na values as 'Not Mentioned'

In [33]:
policedf[policedf['stop_duration'] == '2']

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
55873  1/23/2012     11:17             M          1960.0        52.0   

      driver_race            violation_raw         violation  \
55873       Black  Other Traffic Violation  Moving violation   

       search_conducted      stop_outcome is_arrested stop_duration  \
55873             False  Arrest Passenger        True             2   

       drugs_related_stop  
55873               False

In [34]:
idx = policedf.index[policedf['stop_duration'] == '2']
type(idx)

pandas.core.indexes.numeric.Int64Index

In [35]:
policedf.at[idx, 'stop_duration'] = '2 Min'

In [36]:
policedf[policedf['stop_duration'] == '2 Min']

stop_date stop_time driver_gender  driver_age_raw  driver_age  \
55873  1/23/2012     11:17             M          1960.0        52.0   

      driver_race            violation_raw         violation  \
55873       Black  Other Traffic Violation  Moving violation   

       search_conducted      stop_outcome is_arrested stop_duration  \
55873             False  Arrest Passenger        True         2 Min   

       drugs_related_stop  
55873               False

 We have changed the value '2' to '2 Min' for the stop duration column

In [37]:
policedf['stop_duration'].value_counts()

0-15 Min     47379
16-30 Min    11448
30+ Min       2647
2 Min            1
Name: stop_duration, dtype: int64

In [38]:
policedf['stop_duration'] = policedf['stop_duration'].fillna('Not Mentioned')

We have fillned all na values now no null values will be there in Data frame

In [39]:
policedf.isnull().sum()

stop_date             0
stop_time             0
driver_gender         0
driver_age_raw        0
driver_age            0
driver_race           0
violation_raw         0
violation             0
search_conducted      0
stop_outcome          0
is_arrested           0
stop_duration         0
drugs_related_stop    0
dtype: int64

Counting the genders which were violating the speed limit

In [40]:
policedf[policedf.violation == 'Speeding'].driver_gender.value_counts()

M                25517
F                11686
Not Mentioned        1
Name: driver_gender, dtype: int64

Counting how many time search was conducted on differenr gender

In [41]:
policedf.groupby('driver_gender').search_conducted.sum()

driver_gender
F                 366.0
M                2113.0
Not Mentioned       0.0
Name: search_conducted, dtype: float64

Getting age discription as per each violation

In [42]:
policedf.groupby('violation').driver_age.describe()

count       mean        std   min        25%  \
violation                                                             
Equipment             6516.0  31.686363  11.373176  16.0  23.000000   
Moving violation     11926.0  36.725595  13.231582  15.0  25.000000   
Not Mentioned         4060.0  34.141904   0.348951  18.0  34.148984   
Other                 3583.0  40.178563  12.608331  16.0  30.000000   
Registration/plates   2243.0  32.658692  11.143450  16.0  24.000000   
Seat belt                3.0  30.333333  10.214369  23.0  24.500000   
Speeding             37204.0  33.264582  12.601601  15.0  23.000000   

                           50%        75%   max  
violation                                        
Equipment            28.000000  39.000000  81.0  
Moving violation     35.000000  46.000000  86.0  
Not Mentioned        34.148984  34.148984  38.0  
Other                40.000000  49.500000  86.0  
Registration/plates  30.000000  40.000000  74.0  
Seat belt            26.000000  34.000000  42.0  
Speeding             30.000000  42.000000  88.0